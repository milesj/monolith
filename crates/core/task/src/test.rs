use crate::{FileGroup, Target, Task};
use moon_config::TaskConfig;
use moon_utils::string_vec;
use rustc_hash::FxHashMap;

pub fn create_file_groups_config() -> FxHashMap<String, Vec<String>> {
    let mut map = FxHashMap::default();

    map.insert(
        String::from("static"),
        string_vec![
            "file.ts",
            "dir",
            "dir/other.tsx",
            "dir/subdir",
            "dir/subdir/another.ts",
        ],
    );

    map.insert(String::from("dirs_glob"), string_vec!["**/*"]);

    map.insert(String::from("files_glob"), string_vec!["**/*.{ts,tsx}"]);

    map.insert(String::from("globs"), string_vec!["**/*.{ts,tsx}", "*.js"]);

    map.insert(String::from("no_globs"), string_vec!["config.js"]);

    map
}

pub fn create_file_groups() -> FxHashMap<String, FileGroup> {
    let mut map = FxHashMap::default();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            string_vec![
                "file.ts",
                "dir",
                "dir/other.tsx",
                "dir/subdir",
                "dir/subdir/another.ts",
            ],
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", string_vec!["**/*"]),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", string_vec!["**/*.{ts,tsx}"]),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new("globs", string_vec!["**/*.{ts,tsx}", "*.js"]),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", string_vec!["config.js"]),
    );

    map
}

pub fn create_initial_task(config: Option<TaskConfig>) -> Task {
    Task::from_config(
        Target::new("project", "task").unwrap(),
        &config.unwrap_or_default(),
    )
    .unwrap()
}
