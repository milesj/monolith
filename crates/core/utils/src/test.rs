use crate::path;
use crate::process::output_to_string;
use std::env;
use std::path::{Path, PathBuf};
use std::process::Command;

pub use assert_fs::TempDir;

pub fn run_git_command<P, F>(dir: P, handler: F)
where
    P: AsRef<Path>,
    F: FnOnce(&mut Command),
{
    let mut cmd = Command::new(if cfg!(windows) { "git.exe" } else { "git" });
    cmd.current_dir(dir.as_ref());

    handler(&mut cmd);

    let out = cmd.output().unwrap_or_else(|e| {
        panic!("{:#?}", e);
    });

    if !out.status.success() {
        println!("{}", output_to_string(&out.stdout));
        eprintln!("{}", output_to_string(&out.stderr));
    }
}

pub fn create_sandbox<T: AsRef<str>>(fixture: T) -> assert_fs::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = assert_fs::TempDir::new().unwrap();

    temp_dir
        .copy_from(get_fixtures_dir(fixture), &["**/*"])
        .unwrap();

    temp_dir
        .copy_from(get_fixtures_root(), &["shared-workspace.yml"])
        .unwrap();

    temp_dir
}

pub fn create_sandbox_with_git<T: AsRef<str>>(fixture: T) -> assert_fs::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = create_sandbox(fixture);

    if !temp_dir.path().join(".gitignore").exists() {
        temp_dir
            .child(".gitignore")
            .write_str("node_modules")
            .unwrap();
    }

    // Initialize a git repo so that VCS commands work
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["init", "--initial-branch", "master"]);
    });

    // We must also add the files to the index
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["add", "--all", "."]);
    });

    // And commit them... this seems like a lot of overhead?
    run_git_command(temp_dir.path(), |cmd| {
        cmd.args(["commit", "-m", "Fixtures"])
            .env("GIT_AUTHOR_NAME", "moon tests")
            .env("GIT_AUTHOR_EMAIL", "fakeemail@moonrepo.dev")
            .env("GIT_COMMITTER_NAME", "moon tests")
            .env("GIT_COMMITTER_EMAIL", "fakeemail@moonrepo.dev");
    });

    temp_dir
}

pub fn get_fixtures_root() -> PathBuf {
    let mut root = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    root.push("../../../tests/fixtures");

    path::normalize(root)
}

pub fn get_fixtures_dir<T: AsRef<str>>(dir: T) -> PathBuf {
    get_fixtures_root().join(dir.as_ref())
}

pub fn replace_fixtures_dir<T: AsRef<str>, P: AsRef<Path>>(value: T, dir: P) -> String {
    let dir_str = dir.as_ref().to_str().unwrap();

    // Replace both forward and backward slashes
    value
        .as_ref()
        .replace(dir_str, "<WORKSPACE>")
        .replace(&path::standardize_separators(dir_str), "<WORKSPACE>")
}

pub fn create_moon_command<T: AsRef<Path>>(path: T) -> assert_cmd::Command {
    let mut cmd = assert_cmd::Command::cargo_bin("moon").unwrap();
    cmd.current_dir(path);
    cmd.timeout(std::time::Duration::from_secs(90));
    // Let our code know were running tests
    cmd.env("MOON_TEST", "true");
    // Hide install output as it disrupts testing snapshots
    cmd.env("MOON_TEST_HIDE_INSTALL_OUTPUT", "true");
    // Standardize file system paths for testing snapshots
    cmd.env("MOON_TEST_STANDARDIZE_PATHS", "true");
    // Enable logging for code coverage
    cmd.env("MOON_LOG", "trace");
    cmd
}

pub fn get_assert_output(assert: &assert_cmd::assert::Assert) -> String {
    get_assert_stdout_output(assert) + &get_assert_stderr_output_clean(assert)
}

pub fn get_assert_stderr_output_clean(assert: &assert_cmd::assert::Assert) -> String {
    let mut output = String::new();

    // We need to always show logs for proper code coverage,
    // but this breaks snapshots, and as such, we need to manually
    // filter out log lines and env vars!
    for line in get_assert_stderr_output(assert).split('\n') {
        if !line.starts_with("[error")
            && !line.starts_with("[ warn")
            && !line.starts_with("[ info")
            && !line.starts_with("[debug")
            && !line.starts_with("[trace")
            && !line.starts_with("  MOON_")
            && !line.starts_with("  NODE_")
        {
            output.push_str(line);
            output.push('\n');
        }
    }

    output
}

pub fn get_assert_stderr_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stderr.to_owned()).unwrap()
}

pub fn get_assert_stdout_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stdout.to_owned()).unwrap()
}

pub fn debug_sandbox_files(dir: &Path) {
    for entry in std::fs::read_dir(dir).unwrap() {
        let path = entry.unwrap().path();

        if path.is_dir() {
            debug_sandbox_files(&path);
        } else {
            println!("- {}", path.to_string_lossy());
        }
    }
}

pub fn debug_sandbox(fixture: &assert_fs::TempDir, assert: &assert_cmd::assert::Assert) {
    // List all files in the sandbox
    println!("sandbox:");
    debug_sandbox_files(fixture.path());
    println!("\n");

    // Debug outputs
    println!("stdout:\n{}\n", get_assert_stdout_output(assert));
    println!("stderr:\n{}\n", get_assert_stderr_output(assert));
    println!("status: {:#?}", assert.get_output().status);
}
