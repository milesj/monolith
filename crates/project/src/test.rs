use crate::errors::ProjectError;
use crate::file_group::FileGroup;
use crate::target::Target;
use crate::task::Task;
use crate::token::{TokenResolver, TokenSharedData};
use moon_config::TaskConfig;
use std::collections::HashMap;
use std::path::Path;

pub fn create_file_groups_config() -> HashMap<String, Vec<String>> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        vec![
            "file.ts".to_owned(),
            "dir".to_owned(),
            "dir/other.tsx".to_owned(),
            "dir/subdir".to_owned(),
            "dir/subdir/another.ts".to_owned(),
        ],
    );

    map.insert(String::from("dirs_glob"), vec!["**/*".to_owned()]);

    map.insert(String::from("files_glob"), vec!["**/*.{ts,tsx}".to_owned()]);

    map.insert(
        String::from("globs"),
        vec!["**/*.{ts,tsx}".to_owned(), "*.js".to_owned()],
    );

    map.insert(String::from("no_globs"), vec!["config.js".to_owned()]);

    map
}

pub fn create_file_groups(project_root: &Path) -> HashMap<String, FileGroup> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            vec![
                "file.ts".to_owned(),
                "dir".to_owned(),
                "dir/other.tsx".to_owned(),
                "dir/subdir".to_owned(),
                "dir/subdir/another.ts".to_owned(),
            ],
            project_root,
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", vec!["**/*".to_owned()], project_root),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", vec!["**/*.{ts,tsx}".to_owned()], project_root),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new(
            "globs",
            vec!["**/*.{ts,tsx}".to_owned(), "*.js".to_owned()],
            project_root,
        ),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", vec!["config.js".to_owned()], project_root),
    );

    map
}

pub fn create_expanded_task(
    workspace_root: &Path,
    project_root: &Path,
    config: Option<TaskConfig>,
) -> Result<Task, ProjectError> {
    let mut task = Task::from_config(
        Target::format("project", "task").unwrap(),
        &config.unwrap_or_default(),
    );
    let file_groups = create_file_groups(project_root);
    let metadata = TokenSharedData::new(&file_groups, workspace_root, project_root);

    task.expand_inputs(TokenResolver::for_inputs(&metadata))?;
    task.expand_outputs(TokenResolver::for_outputs(&metadata))?;
    task.expand_args(TokenResolver::for_args(&metadata))?; // Must be last

    Ok(task)
}
