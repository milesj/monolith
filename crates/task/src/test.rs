use crate::{FileGroup, Target, Task, TaskError, TokenResolver, TokenSharedData};
use moon_config::{ProjectConfig, TaskConfig};
use moon_utils::string_vec;
use std::collections::HashMap;
use std::path::Path;

pub fn create_file_groups_config() -> HashMap<String, Vec<String>> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        string_vec![
            "file.ts",
            "dir",
            "dir/other.tsx",
            "dir/subdir",
            "dir/subdir/another.ts",
        ],
    );

    map.insert(String::from("dirs_glob"), string_vec!["**/*"]);

    map.insert(String::from("files_glob"), string_vec!["**/*.{ts,tsx}"]);

    map.insert(String::from("globs"), string_vec!["**/*.{ts,tsx}", "*.js"]);

    map.insert(String::from("no_globs"), string_vec!["config.js"]);

    map
}

pub fn create_file_groups() -> HashMap<String, FileGroup> {
    let mut map = HashMap::new();

    map.insert(
        String::from("static"),
        FileGroup::new(
            "static",
            string_vec![
                "file.ts",
                "dir",
                "dir/other.tsx",
                "dir/subdir",
                "dir/subdir/another.ts",
            ],
        ),
    );

    map.insert(
        String::from("dirs_glob"),
        FileGroup::new("dirs_glob", string_vec!["**/*"]),
    );

    map.insert(
        String::from("files_glob"),
        FileGroup::new("files_glob", string_vec!["**/*.{ts,tsx}"]),
    );

    map.insert(
        String::from("globs"),
        FileGroup::new("globs", string_vec!["**/*.{ts,tsx}", "*.js"]),
    );

    map.insert(
        String::from("no_globs"),
        FileGroup::new("no_globs", string_vec!["config.js"]),
    );

    map
}

pub fn create_initial_task(config: Option<TaskConfig>) -> Task {
    Task::from_config(
        Target::format("project", "task").unwrap(),
        &config.unwrap_or_default(),
    )
}

pub fn create_expanded_task(
    workspace_root: &Path,
    project_root: &Path,
    config: Option<TaskConfig>,
) -> Result<Task, TaskError> {
    let mut task = create_initial_task(config);
    let file_groups = create_file_groups();
    let project_config = ProjectConfig::new(project_root);
    let metadata =
        TokenSharedData::new(&file_groups, workspace_root, project_root, &project_config);

    task.expand_env(&metadata)?;
    task.expand_inputs(TokenResolver::for_inputs(&metadata))?;
    task.expand_outputs(TokenResolver::for_outputs(&metadata))?;
    task.expand_args(TokenResolver::for_args(&metadata))?; // Must be last

    Ok(task)
}
