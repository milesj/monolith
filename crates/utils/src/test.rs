use crate::path;
use std::env;
use std::path::{Path, PathBuf};

pub fn create_fixtures_sandbox(dir: &str) -> assert_fs::fixture::TempDir {
    use assert_fs::prelude::*;

    let temp_dir = assert_fs::fixture::TempDir::new().unwrap();

    temp_dir
        .copy_from(get_fixtures_dir(dir), &["**/*"])
        .unwrap();

    temp_dir
}

pub fn get_fixtures_dir(dir: &str) -> PathBuf {
    get_fixtures_root().join(dir)
}

pub fn get_fixtures_root() -> PathBuf {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("../../tests/fixtures");

    path.canonicalize().unwrap()
}

pub fn replace_fixtures_dir(value: &str, dir: &Path) -> String {
    value.replace(dir.to_str().unwrap(), "<WORKSPACE>")
}

// We need to do this so slashes are accurate and always forward
pub fn wrap_glob(path: &Path) -> PathBuf {
    PathBuf::from(path::normalize_glob(path))
}

pub fn create_moon_command(fixture: &str) -> assert_cmd::Command {
    let mut cmd = create_moon_command_in(&get_fixtures_dir(fixture));
    // Never cache in these tests since they're not in a sandbox
    cmd.env("MOON_CACHE", "off");
    cmd
}

pub fn create_moon_command_in(path: &Path) -> assert_cmd::Command {
    let mut cmd = assert_cmd::Command::cargo_bin("moon").unwrap();
    cmd.current_dir(path);
    // Let our code know were running tests
    cmd.env("MOON_TEST", "true");
    // Hide install output as it disrupts testing snapshots
    cmd.env("MOON_TEST_HIDE_INSTALL_OUTPUT", "true");
    // Standardize file system paths for testing snapshots
    cmd.env("MOON_TEST_STANDARDIZE_PATHS", "true");
    // Uncomment for debugging
    // cmd.arg("--logLevel");
    // cmd.arg("trace");
    cmd
}

pub fn get_assert_output(assert: &assert_cmd::assert::Assert) -> String {
    get_assert_stdout_output(assert) + &get_assert_stderr_output(assert)
}

pub fn get_assert_stderr_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stderr.to_owned()).unwrap()
}

pub fn get_assert_stdout_output(assert: &assert_cmd::assert::Assert) -> String {
    String::from_utf8(assert.get_output().stdout.to_owned()).unwrap()
}
