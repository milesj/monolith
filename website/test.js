const readline = require('readline');

console.log('START');

const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
});

function ask(question) {
	console.log('ASK');
	rl.question(question, (answer) => {
		console.log('Q');
		rl.write(`The answer received:  ${answer}\n`);

		ask(question);
	});
}

ask('What is your name: ');
