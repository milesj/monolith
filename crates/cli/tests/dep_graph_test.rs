use insta::assert_snapshot;
use moon_utils::test::{create_moon_command, get_assert_output};

#[test]
fn all_by_default() {
    let assert = create_moon_command("cases").arg("dep-graph").assert();
    let dot = get_assert_output(&assert);

    // Snapshot is not deterministic
    assert_eq!(dot.split("\n").count(), 227);
}

#[test]
fn focused_by_target() {
    let assert = create_moon_command("cases")
        .arg("dep-graph")
        .arg("node:standard")
        .assert();

    assert_snapshot!(get_assert_output(&assert));
}

#[test]
fn includes_dependencies_when_focused() {
    let assert = create_moon_command("cases")
        .arg("dep-graph")
        .arg("dependsOn:standard")
        .assert();

    assert_snapshot!(get_assert_output(&assert));
}

#[test]
fn includes_dependents_when_focused() {
    let assert = create_moon_command("cases")
        .arg("dep-graph")
        .arg("depsC:standard")
        .assert();

    assert_snapshot!(get_assert_output(&assert));
}
