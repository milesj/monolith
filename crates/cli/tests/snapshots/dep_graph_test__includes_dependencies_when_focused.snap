---
source: crates/cli/tests/dep_graph_test.rs
assertion_line: 30
expression: get_assert_output(&assert)
---
digraph {
    0 [ label="SetupToolchain" style=filled, shape=oval, fillcolor=black, fontcolor=white]
    1 [ label="InstallNodeDeps" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    2 [ label="SyncNodeProject(dependsOn)" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    3 [ label="SyncNodeProject(depsC)" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    4 [ label="SyncNodeProject(depsB)" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    5 [ label="SyncNodeProject(depsA)" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    6 [ label="RunTarget(dependsOn:standard)" style=filled, shape=oval, fillcolor=gray, fontcolor=black]
    1 -> 0 [ arrowhead=box, arrowtail=box]
    2 -> 0 [ arrowhead=box, arrowtail=box]
    3 -> 0 [ arrowhead=box, arrowtail=box]
    2 -> 3 [ arrowhead=box, arrowtail=box]
    4 -> 0 [ arrowhead=box, arrowtail=box]
    2 -> 4 [ arrowhead=box, arrowtail=box]
    5 -> 0 [ arrowhead=box, arrowtail=box]
    2 -> 5 [ arrowhead=box, arrowtail=box]
    6 -> 1 [ arrowhead=box, arrowtail=box]
    6 -> 2 [ arrowhead=box, arrowtail=box]
}



